### ISSUE TITLE
Concern regarding dropping streambuf

### ISSUE THREAD

**Issue by heilkn** [Lyberta/cpp-io#4]
Hello, 

first of all, thank you for your hard work. I have found this document very well written and agree with almost of your arguments and proposals. However, one short comment left me wondering. I would like to ask you to reconsider dropping the indirection provided by streambuf. Or alternatively, proposing a good alternative.

I have found that the aforementioned indirection allows composition of stream buffers. One can easily wrap one buffer into another and add CRC-verification, compression or en-/decryption transparently into a stream.

Is something like this possible with your proposal and reference implementation and I simply have missed the point?

Once again, thank you very much for your work. I am looking forward reading from you.

Kind regards

Konstantin

**Comment by Lyberta** [comment]
> I have found that the aforementioned indirection allows composition of stream buffers.

Yes, but since this proposal uses templates, you can provide your own stream class that does composition.

In fact, there are 2 such streams in the reference implementation already: [`BufferedFile`](https://github.com/Lyberta/cpp-io-impl/blob/master/Headers/Internal/buffered_file.h) and [`SynchronizedStream`](https://github.com/Lyberta/cpp-io-impl/blob/master/Headers/Internal/synchronized_stream.h). First one [implements](https://github.com/Lyberta/cpp-io-impl/blob/master/Sources/buffered_file.cpp) buffered file IO in portable way and second one [implements](https://github.com/Lyberta/cpp-io-impl/blob/master/Headers/Internal/synchronized_stream.hpp) thread safety required for standard stream objects.
